conclave-host/src/main/kotlin/com/r3/conclave/host/internal/GramineEnclaveHandle.kt

```diff
@@ -46,24 +46,22 @@ class GramineEnclaveHandle(
     }
 
     override fun initialise() {
-        /** Start the local call interface. */
-        val interfaceStartTask = FutureTask { enclaveInterface.start() }
-        val interfaceStartThread = Thread(interfaceStartTask).apply { start() }
-
         try {
+            /** Bind a port for the interface to use. */
+            val port = enclaveInterface.bindPort()
+
             /**
              * Start the enclave process, passing the port that the call interface is listening on.
              * TODO: Implement a *secure* method for passing port to the enclave.
              */
             processGramineDirect = ProcessBuilder()
                 .inheritIO()
                 .directory(enclaveDirectory.toFile())
-                .command("gramine-direct", "java", "-cp", GRAMINE_ENCLAVE_JAR_NAME, "com.r3.conclave.enclave.internal.GramineEntryPoint", enclaveInterface.port.toString())
+                .command("gramine-direct", "java", "-cp", GRAMINE_ENCLAVE_JAR_NAME, "com.r3.conclave.enclave.internal.GramineEntryPoint", port.toString())
                 .start()
 
             /** Wait for the local call interface start process to complete. */
-            interfaceStartThread.join()     // wait for start process to finish
-            interfaceStartTask.get()        // throw if start failed
+            enclaveInterface.start()
```

Comment:
This is much better. I think we can improve it even further by removing the synchronized blocks in `start` and `close`.

Reply:
Ehhh, not sure about this.
Maybe in start if we assume that no calls will arrive until `start()` has returned. Definitely not in `close()` though, as there's a race condition with the state check in `initiateOutgoingCall()`.
More generally I think the state manager could use a little work. With any state machine there is a process that takes it from one state to another. I think this could be better expressed by passing a lambda to the transition methods that contains the actions that need to accompany the state change. There's also possibly an error condition if the state transition fails. I also think all state transitions should all be atomic. It really doesn't make sense for a state machine to be operated on by multiple threads anyway.

Reply:
Close has to wait for currently running calls to finish. There is a race condition between changing the state to stopped and acquiring the concurrent call semaphore that could result in a thread (or threads) waiting forever in `initiateOutgoingCall()` otherwise.